### PR TITLE
[FIX] mail: multi recipient duplication (16.0)

### PR DESCRIPTION
**Pull request by Julien00859**
Install mass-mailing and crm. Change the email address of Brandon to a@example.com;b@example.com. In the list view of crm, add the select all leads and change their customer to Brandon (the customer column is not displayed by default, just make it visible). Create a new mass-mailing with the recipient list as "Lead/Opportunity". Start the campaign. Brandon receives as many emails as there are leads but he shall only receive one.

The system has a known limitation when it comes to filtering duplicates: it skips all records that have multiple recipients. In this case Brandon has two: [a@example.com](mailto:a@example.com) and [b@example.com](mailto:b@example.com). The de-duplication mechanism was skipped for every lead he was the customer of and each time a new email was sent, spamming him.

In this work we make it possible to also process records with multiple recipients. It is a best-effort and will still let some duplicates through. Nonetheless it solves the current problem with minimal changes.

Note: any([]) and any(['']) are both False while all([]) is True, hence we now check for empty list / empty email first otherwise an empty list would be considered to be opt-out instead of empty.

Task-3927361

### CODE REVIEW

File: addons/mail/wizard/mail_compose_message.py [hunk 1]
```diff
@@ -518,21 +510,25 @@ def _process_state(self, mail_values_dict):
...
-            elif optout_emails and mail_to in optout_emails:
+            elif optout_emails and all(
+                    mail_to in optout_emails for mail_to in recipients['mail_to']
+                ):
                 mail_values['state'] = 'cancel'
                 mail_values['failure_type'] = 'mail_optout'
-            elif done_emails and mail_to in done_emails and not mailing_document_based:
+            elif done_emails and not mailing_document_based and all(
+                    mail_to in done_emails for mail_to in recipients['mail_to']
+                ):
```
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_compose_message.py]: Would say that any email in blocklist win, but that's more a personal choice. Not sure if this was discussed ?

File: addons/mail/wizard/mail_compose_message.py [hunk 2]
```diff
@@ -518,21 +510,25 @@ def _process_state(self, mail_values_dict):
...
+            # void or falsy values -> error
+            elif not any(recipients['mail_to']):
                 mail_values['state'] = 'cancel'
                 mail_values['failure_type'] = 'mail_email_missing'
-            elif not mail_to_normalized:
+            elif not any(recipients['mail_to_normalized']):
                 mail_values['state'] = 'cancel'
                 mail_values['failure_type'] = 'mail_email_invalid'
-            elif optout_emails and mail_to in optout_emails:
+            elif optout_emails and all(
```
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_compose_message.py:520]: Same for optout, but less sure about that one.

File: addons/test_mail/tests/test_mail_composer.py [hunk 3]
```diff
@@ -2028,12 +2028,12 @@ def test_mail_composer_wtpl_recipients_email_fields(self):
         #   there are more partners than email to notify;
         self.assertEqual(len(self._new_mails), 2, 'Should have created 2 mail.mail')
         self.assertEqual(
-            len(self._mails), (len(new_partners) + 2) * 2,
-            f'Should have sent {(len(new_partners) + 2) * 2} emails, one / recipient ({len(new_partners)} mailed partners + partner_1 + partner_2) * 2 records')
-        for record in self.test_records:
+            len(self._mails), len(new_partners) + 2,
+            f'Should have sent {len(new_partners) + 2} emails, one / recipient ({len(new_partners)} mailed partners + partner_1 + partner_2)')
+        for record, status in zip(self.test_records, ('sent', 'cancel')):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:2033]: Why is the second one canceled ? Is the content exactly the same ? In that case would probably update template content, as purpose here is to really test a multi send, not really duplicate check, would prefer keeping "sent" emails.

File: addons/test_mail_full/tests/test_mass_mailing.py [hunk 4]
```diff
@@ -57,6 +68,11 @@ def test_mailing_w_blacklist_opt_out(self):
             elif recipient == recipient_dup_1:
                 recipient_info['trace_status'] = "cancel"
                 recipient_info['failure_type'] = "mail_dup"
+            elif recipient == recipient_dup_2:
+                pass
```
Comment (tde-banana-odoo) [inline addons/test_mail_full/tests/test_mass_mailing.py:72]: Why pass ? Can't we go in check below ?

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for retaking it and fixing it a bit in-depth ! Few comments on my side.

File: addons/mail/wizard/mail_compose_message.py (hunk 1 above)
Comment (Julien00859) [reply addons/mail/wizard/mail_compose_message.py]: It has not been discussed, my opinion is that customers can still create a filtrer in their own mailbox if they *really* don't want to receive those emails, but the opposite is impossible. It also is an improvement over the current system, that just send all emails when there are multiple recipients. Changing the behavior for `any` would be more disruptive, why not in master, not sure in stable.

File: addons/mail/wizard/mail_compose_message.py (hunk 2 above)
Comment (Julien00859) [reply addons/mail/wizard/mail_compose_message.py:520]: Same a above

File: addons/test_mail_full/tests/test_mass_mailing.py (hunk 4 above)
Comment (Julien00859) [reply addons/test_mail_full/tests/test_mass_mailing.py:72]: I don't know, some check in the `else` block were failing and I didn't understand those...

Make it work -> make it good (-> make it fast)

I made it work, would you help me make it good? :D

File: addons/test_mail/tests/test_mail_composer.py (hunk 3 above)
Comment (Julien00859) [reply addons/test_mail/tests/test_mail_composer.py:2033]: It is sending the same template to both test_records, the second one is cancelled for mail_dup. I've tried for 2 hours to find a solution but I achieved nothing. Do you have a suggestion?

Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py:2033]: In 16, duplication is done only on email (not on actual content like in newer versions, where we check attachments, subject, ...). You can use the ``mailing_document_based`` context key to remove the duplicate check in this case (as we are not trying to check duplicate check here, more email parsing). It was added in stable to indicate there are documents where same email does not mean duplicate (e.g. invoice, sale orders, ...).

File: addons/test_mail_full/tests/test_mass_mailing.py (hunk 4 above)
Comment (tde-banana-odoo) [reply addons/test_mail_full/tests/test_mass_mailing.py:72]: What was failing :D ?

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

Woop woop !

**Pull request closed without merging**